### PR TITLE
[network-addons] introduce expo-network-addons

### PR DESCRIPTION
**Pull request by Kudo**
# Why

introduce expo-network-addons which is an opt-in networking addons. currently only support one feature - add brotli compression support for all okhttp requests on android.
close ENG-538

# How

- add expo-network-addons
  - with an autolinked gradle-plugin to do bytecode manipulation for the `ExpoOkHttpInterceptor`, currently only use `BrotliInterceptor` under the hood.
  - add proguard rules to keep the java names we want to manipulate

### compatibility between `expo-dev-launcher` and `expo-network-addons`

since we cannot do bytecode manipulation twice, to deal with the compatibility issues between expo-dev-launcher and expo-network-addons. if we find both library are installed, we will let `expo-network-addons` take the ownership and install dev-launcher's bytecode.

Screenshot 2023-06-20 at 11 09 50 PM

# Test Plan

- bare-expo `fetch('https://www.cloudflare.com')` and use mitmproxy to check brotli headers
- test bare-expo on release build + `enableProguardInReleaseBuilds = true`
- decompile the dex and check if the code is expected
Screenshot 2023-06-20 at 9 43 40 PM

# Checklist

- n/a Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
  no user faced changes.
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by Kudo** [comment]
android unit tests are broken on main 😑

File: packages/expo-network-addons/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+### Installation
+
+```
+npx expo install expo-network-addons
+```
+
+## Contributing
+
+Contributions are very welcome! Please refer to guidelines described in the [contributing guide][../../contributing.md].
```
Comment (gabrieldonadel) [inline packages/expo-network-addons/README.md:24]: Nit
```suggestion
Contributions are very welcome! Please refer to guidelines described in the [contributing guide](../../contributing.md).
```

**Comment by brentvatne** [comment]
CI should be fixed in https://github.com/expo/expo/commit/884edb9cba98934d034f35aaf9d61a375d6c96c6

**Review by ide — COMMENTED** [review]
I will review the NetworkAddonsPlugin more closely but wanted to bring up two discussion topics:

1. Do you expect we will have more plugins that need to rewrite bytecode? I'm asking because the dev client plugin is now aware of the network addons plugin, which is easy to maintain if there are just two plugins that rewrite bytecode. But if there are more plugins that rewrite bytecode, the complexity becomes O(n^2).
2. Do you think the `expo-network-addons` package will become too big as it gains more features in addition to Brotli support? Will developers need to include extra addons even if they don't use them? Should this be `expo-network-brotli`, or is that too complicated (especially because of (1) above)?

**Comment by Kudo** [comment]
> 1. Do you expect we will have more plugins that need to rewrite bytecode? I'm asking because the dev client plugin is now aware of the network addons plugin, which is easy to maintain if there are just two plugins that rewrite bytecode. But if there are more plugins that rewrite bytecode, the complexity becomes O(n^2).

it is because they rewrite bytecode to the same method `OkHttpClient::Builder.build()`. i don't think we will have another plugin to aim at the method. if that happens, we may have to unify the bytecode manipulation just in single place, i.e. we may then add expo-network-addons as a dependency from expo-dev-launcher or any other new packages for bytecode manipulation.

> 2. Do you think the `expo-network-addons` package will become too big as it gains more features in addition to Brotli support? Will developers need to include extra addons even if they don't use them? Should this be `expo-network-brotli`, or is that too complicated (especially because of (1) above)?

since last time we've discussed [cronet](https://github.com/google/cronet-transport-for-okhttp), e.g. for QUIC/HTTP3 support, that's something we can add to this package. that's why proposing the package name as "network-addons". we can use gradle.properties/Podfile.properties.json to control the feature toggling and only include extra native dependencies if necessary.

**Review by ide — APPROVED** [review]
The Proguard rules and NetworkAddonsPlugin depend on several of the class names. If the class names change, the code will break. It would be good to add comments to these classes saying it is necessary to update the plugin and Proguard rules if the classes are renamed since the static type checker will not catch these errors.

**Review by wschurman — APPROVED** [review]

File: packages/expo-dev-launcher/expo-dev-launcher-gradle-plugin/src/main/kotlin/expo/modules/devlauncher/DevLauncherPlugin.kt [hunk 2]
```diff
@@ -70,6 +78,10 @@ abstract class DevLauncherPlugin : Plugin<Project> {
 
   class OkHttpClientCustomBuildMethod(api: Int, methodVisitor: MethodVisitor) : MethodVisitor(api, methodVisitor) {
     override fun visitCode() {
+      //
+      // NOTE: The following code should be in sync with **packages/expo-network-addons/expo-network-addons-gradle-plugin/src/main/kotlin/expo/modules/networkaddons/NetworkAddonsPlugin.kt**
```
Comment (ide) [inline packages/expo-dev-launcher/expo-dev-launcher-gradle-plugin/src/main/kotlin/expo/modules/devlauncher/DevLauncherPlugin.kt:82]:
```suggestion
      // NOTE: The following code should be kept in sync with **packages/expo-network-addons/expo-network-addons-gradle-plugin/src/main/kotlin/expo/modules/networkaddons/NetworkAddonsPlugin.kt**
```

File: packages/expo-network-addons/expo-network-addons-gradle-plugin/src/main/kotlin/expo/modules/networkaddons/NetworkAddonsPlugin.kt [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+      visitVarInsn(Opcodes.ALOAD, 0)
+      visitTypeInsn(Opcodes.NEW, "expo/modules/networkaddons/ExpoOkHttpInterceptor")
+      visitInsn(Opcodes.DUP)
+      visitMethodInsn(Opcodes.INVOKESPECIAL, "expo/modules/networkaddons/ExpoOkHttpInterceptor", "<init>", "()V", false)
+      visitTypeInsn(Opcodes.CHECKCAST, "okhttp3/Interceptor")
+      visitMethodInsn(Opcodes.INVOKEVIRTUAL, "okhttp3/OkHttpClient\$Builder", "addInterceptor", "(Lokhttp3/Interceptor;)Lokhttp3/OkHttpClient\$Builder;", false)
+
+      if (parameters.debugVariant.getOrElse(false) && parameters.devLauncherInstalled.getOrElse(false)) {
+        //
+        // NOTE: The following code should be in sync with **packages/expo-dev-launcher/expo-dev-launcher-gradle-plugin/src/main/kotlin/expo/modules/devlauncher/DevLauncherPlugin.kt**
```
Comment (ide) [inline packages/expo-network-addons/expo-network-addons-gradle-plugin/src/main/kotlin/expo/modules/networkaddons/NetworkAddonsPlugin.kt:101]:
```suggestion
        // NOTE: The following code should be kept in sync with **packages/expo-dev-launcher/expo-dev-launcher-gradle-plugin/src/main/kotlin/expo/modules/devlauncher/DevLauncherPlugin.kt**
```

File: packages/expo-dev-launcher/expo-dev-launcher-gradle-plugin/src/main/kotlin/expo/modules/devlauncher/DevLauncherPlugin.kt [hunk 4]
```diff
@@ -24,9 +24,17 @@ abstract class DevLauncherPlugin : Plugin<Project> {
   override fun apply(project: Project) {
     val enableNetworkInspector = project.properties["EX_DEV_CLIENT_NETWORK_INSPECTOR"]?.toString()?.toBoolean()
     if (enableNetworkInspector != null && enableNetworkInspector) {
+      // When expo-network-addons is installed, we will let it to do the bytecode manipulation
+      val networkAddonsInstalled = project.findProject(":expo-network-addons") != null
+      if (networkAddonsInstalled) {
+        logger.warn("[DevLauncherPlugin] expo-network-addons is installed and will take the plugin's ownership")
```
[the remark this answers is not in the document]
Comment (ide) [reply packages/expo-dev-launcher/expo-dev-launcher-gradle-plugin/src/main/kotlin/expo/modules/devlauncher/DevLauncherPlugin.kt:30]: Requesting these changes still

**Review by ide — APPROVED** [review]

**Pull request merged**